### ISSUE TITLE
salt-cloud GoGrid exception when using map file

### ISSUE THREAD

**Issue by luciddr34m3r** [saltstack/salt#14666]
After attempting to launch GoGrid minions in a map file, the first minion installs properly, but an exception prevents the script from continuing. I get the following error: https://gist.github.com/luciddr34m3r/1364babefaa344f0508c

I added a debug statement (the info log line directly above the error), which prints the value of the 'data' variable at the time of the exception on the following line: 
https://github.com/saltstack/salt/blob/develop/salt/cloud/clouds/gogrid.py#L260

update() expects a dictionary, but 'data' at that time is actually a Node object. This seems to be the case if the node creation succeeds, as seen here, or fails, as seen in https://github.com/saltstack/salt/issues/13365 

Not sure if the two issues are directly related, but I was encouraged to raise a new issue for this.

**Comment by rallytime** [comment]
@luciddr34m3r I know it's been a while since this issue was filed, but @techhat recently converted the current GoGrid driver from depending on apache-libcloud to using GoGrid's REST API. This fixed my issue that you linked to above with being able to provision VMs (#13365), and I bet this issue is fixed, too.

Are you in a position to give the refactored driver a try? The relevant pull request is here: https://github.com/saltstack/salt/pull/24634. Or you can grab the driver off of the develop branch: https://github.com/saltstack/salt/blob/develop/salt/cloud/clouds/gogrid.py

**Comment by luciddr34m3r** [comment]
Oh geez. Um, we actually dont have a test environment in GoGrid and we are
using 100% of our IP's for production. I can check with mydevops guy
tomorrow but I'm not sure if I can test it right now. Thanks for letting me
know. If we can, I'll hit you up.

**Comment by rallytime** [comment]
@luciddr34m3r No problem. I can probably try this in the next day or so as well, so no worries at all.

**Comment by luciddr34m3r** [comment]
Thanks for the update! We do have GoGrid infrastructure managed in salt
still, but it's really stable at the moment,.

**Comment by rallytime** [comment]
@luciddr34m3r I just tested this and it looks like it isn't quite there yet, though with a different error.

@techhat Mind taking a look at this? When you use a map file with GoGrid, one of the  machines gets provisioned just fine, but the other does not. The following stacktrace occurs, and is sad from there:
```
[ERROR   ] Error creating nt-map-test-1 on GOGRID

The following exception was thrown when trying to run the initial deployment:
Traceback (most recent call last):
  File "/root/SaltStack/salt/salt/cloud/clouds/gogrid.py", line 107, in create
    data = _query('grid', 'server/add', args=create_kwargs)
  File "/root/SaltStack/salt/salt/cloud/clouds/gogrid.py", line 542, in _query
    return result['dict']
KeyError: 'dict'
```

**Comment by luciddr34m3r** [comment]
Does this error happen in parallel and in serial? Gogrid doesn't seem to do
so great with asynchronous calls

**Comment by rallytime** [comment]
@luciddr34m3r Ah, good catch. That error only happens in parallel. I tried it this morning without `-P` and it worked just fine.

**Comment by luciddr34m3r** [comment]
I believe this to actually be an issue with GoGrid's provisioning system.
Someone should actually probably advise them to change their API. That
being said, if you are deploying a map file that has GoGrid and AWS both,
it means you cannot deploy in parallel. We use a non-homogeneous makeup,
and I had to create different map files for parallel-safe and
parallel-unsafe providers in order to get things up quickly. GoGrid had the
slowest deploy process by far.

**Comment by rallytime** [comment]
@luciddr34m3r Yeah I have noticed that deploying to GoGrid is exceptionally slow as well. Thanks for the useful information here. I think the best resolution for this is to document this both in the [cloud map documentation](https://docs.saltstack.com/en/latest/topics/cloud/map.html) as well as in the go-grid specific docs. I'll see if I can add something there today.

**Comment by rallytime** [comment]
@luciddr34m3r I added some notes/warnings to all of the places I thought were relevant for the GoGrid and `-P` combination not working on mapfiles. 

Other than that, I think at this point nothing else can be done about this bug. Do you think it is ok to close?
